core/src/main/scala/org/http4s/blaze/channel/nio1/NIO1HeadStage.scala

```diff
@@ -214,22 +232,37 @@ private[nio1] final class NIO1HeadStage(
   final override def writeRequest(data: Seq[ByteBuffer]): Future[Unit] = {
     logger.trace(s"NIO1HeadStage Write Request: $data")
     val p = Promise[Unit]
-    selectorLoop.executeTask(new Runnable {
-      override def run(): Unit =
+    selectorLoop.executeTask(new selectorLoop.LoopRunnable {
+      override def run(scratch: ByteBuffer): Unit =
         if (closedReason != null) {
           p.failure(closedReason)
           ()
         } else if (writePromise == null) {
           val writes = data.toArray
-          if (!BufferTools.checkEmpty(writes)) { // Non-empty buffer
-            writePromise = p
-            writeData = writes
-            setOp(SelectionKey.OP_WRITE)
-            p.future
-            ()
+          if (!BufferTools.checkEmpty(writes)) {
+            // Non-empty buffers. First we check to see if we can immediately
+            // write all the data to save a trip through the NIO event system.
+            performWrite(ch, scratch, writes) match {
```

Comment:
I like this. All small HTTP responses can be written directly without NIO round trip.